homeassistant/components/recorder/migration.py

```diff
@@ -507,6 +507,18 @@ def _apply_update(engine, session, new_version, old_version):  # noqa: C901
             "statistics",
             ["sum_increase DOUBLE PRECISION"],
         )
+        # Try to change the character set of the statistic_meta table
+        if engine.dialect.name == "mysql":
+            for table in ["events", "states"]:
+                try:
+                    connection.execute(
+                        text(
```

Comment:
Should we check if the table already has the right charset/collation before doing the alter?

Not sure how expensive this operation will be for a large database.

Reply:
I would assume it's a no-op if the character set is already correct?

Reply:
I hope its a no-op, but I've seen other cases with MySQL where I expected it to do nothing, but it rebuilt the whole table unexpectedly. I guess we have to test this to know.

Reply:
Looks like it is a no-op

```
MariaDB [homeassistant]> ALTER TABLE states CONVERT TO CHARACTER SET utf8mb3 COLLATE utf8mb3_unicode_ci;
Query OK, 71284 rows affected (23.267 sec)             
Records: 71284  Duplicates: 0  Warnings: 0

MariaDB [homeassistant]> ALTER TABLE states CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
Query OK, 71285 rows affected (12.383 sec)             
Records: 71285  Duplicates: 0  Warnings: 0

MariaDB [homeassistant]> ALTER TABLE states CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
Query OK, 0 rows affected (0.090 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [homeassistant]> 
```